### ISSUE TITLE
[SignalR] Multiple simultaneous disconnects render SignalR unresponsive for an extended period of time

### ISSUE THREAD

**Issue by sikri-eic** [dotnet/aspnetcore#27246]
### Describe the bug
I have a SignalR test server with 5 hubs and a C# test client. Via the client, I create around 5000 connections towards each hub of the server (5x5000). Then stop all connections "at the same time" (via `await connection.StopAsync()`). While handling the disconnects, SignalR server seems to be unresponsive for an extended period of time (it does not respond to simple controller actions sometimes up to a minute). It looks like it is busy handling these exceptions:

- `Microsoft.AspNetCore.Connections.ConnectionResetException: 'The client has disconnected'`
- `System.InvalidOperationException: 'Reading is not allowed after reader was completed.'`
- `System.Net.WebSockets.WebSocketException: 'The remote party closed the WebSocket connection without completing the close handshake.'`

This only happens when I create some "service" classes that use the hubs via `IHubContext`. 

### To Reproduce
I think I managed to create the minimal setup for this: 1 server, 1 client and a Powershell script to monitor the server.

#### Server
Code for my test server can be found here: https://github.com/sikri-eic/signalrserverdisconnectissue
It has 5 hubs and 3 "service" classes that use 3 of the hubs via `IHubContext`

#### Client
Code for my test client can be found here: https://github.com/sikri-eic/signalrservertestclient

Please pay attention to `ServerUrl` definition and update it according to your setup.

#### Monitoring script

I use the following PowerShell script to monitor the Ping controller so that I know the server is responding:

```PS
while ($true) {
	Start-Sleep -s 1
	try {
		# Pay attention to the URL below
		$timeout = Measure-Command { Invoke-WebRequest https://localhost:44362/ping }
		$seconds = [int]($timeout.Seconds)
		write-host -NoNewline "$seconds|"
	} catch {
		$_.Exception.Response.StatusCode.Value__
		exit
	}
}
```
Perhaps I'm doing something wrong?

**Edit:** I suspected the way I implemented PingController and separated the calls to the service classes to another controller and called that from a different script, but that did not help.

### Exceptions (if any)
- `Microsoft.AspNetCore.Connections.ConnectionResetException: 'The client has disconnected'`
- `System.InvalidOperationException: 'Reading is not allowed after reader was completed.'`
- `System.Net.WebSockets.WebSocketException: 'The remote party closed the WebSocket connection without completing the close handshake.'`

### Further technical details
- ASP.NET Core version: **3.1**
- Include the output of `dotnet --info`
```
.NET Core SDK (reflecting any global.json):
 Version:   3.1.402
 Commit:    9b5de826fd

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.18363
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\3.1.402\

Host (useful for support):
  Version: 3.1.8
  Commit:  9c1330dedd

.NET Core SDKs installed:
  3.1.401 [C:\Program Files\dotnet\sdk]
  3.1.402 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.22 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.22 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.22 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.8 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```
- The IDE (VS / VS Code/ VS4Mac) you're running on, and it's version: VS 16.7.4

**Comment by BrennanConroy** [comment]
I don't see an issue unless I turn on debug logs in the console. That would make sense because console logging is a slow operation, and if you log too much at once you'll block the server until the console output backlog decreases below a threshold.

If you're seeing this without console logging could you please explain how to repro it?

**Comment by sikri-eic** [comment]
@BrennanConroy, thank you for the prompt response. You are right. When I start the SignalR test server without debugging, I cannot reproduce the problem.

I created this bug because we have a problem in one of our deployments (where logging in debug mode cannot possibly be a problem). I tried to create a minimal project that would reproduce the issue, and apparently, I failed :/ (I didn't consider debug console).

I will investigate this a bit more and get back soon (hopefully early next week).
